Speed up Dictionary and DICT_FSST compression by replacing std::unordered_map with PrimitiveDictionary

yan-alex opened the pull request:
# Goal
[#6379](https://github.com/duckdblabs/duckdb-internal/issues/6379) This PR replaces the usage of std::unordered_map by our own implementation of a linear probing hash table to speed up compression.

# Benchmarking


## storage persistent v1.2.0 force_compression='dictionary'
Now we are actually running with DICT_FSST by setting `storage persistent v1.4.0` in the benchmarks, the results are better:
<img width="1038" height="600" alt="newplot (11)" src="https://github.com/user-attachments/assets/84862dc3-ca08-4dd5-9a99-ccd3462619a7" />

## storage persistent v1.2.0 force_compression='dictionary'
I ran experiments with allocation size of `max_unique_count_across_all_segments` and `max_unique_count_across_all_segments * 2` for the dictionary, here are the results:

### max_unique_count_across_all_segments

<img width="1038" height="600" alt="newplot (12)" src="https://github.com/user-attachments/assets/c0d8c92b-bf5f-4221-8bf0-0ccdba07d142" />

### max_unique_count_across_all_segments*2

<img width="1038" height="600" alt="newplot (11)" src="https://github.com/user-attachments/assets/fd9ac723-3688-4a27-ae95-6d2326686a80" />


There is a slight improvement, which is why I kept the *2 allocation.

Commit:
Dictionary compression using PrimitiveDictionary

Commit:
format

Commit:
Custom Clear() method for PrimitiveDictionary

Commit:
entry.index + 1

Commit:
make format-fix

Commit:
DictFSST with PrimitiveDictionary

Commit:
.

Commit:
Insert() instead of Replace()

Commit:
dict_fsst index + 1

Commit:
Pass exact maximum_target_capacity_p to PrimitiveDictionary, for optimal allocation

Commit:
Fix DictionaryCompressionStorage::InitCompression signature, format

Commit:
Keep track of max_unique_count_across_all_segments to allocate dictionary

Commit:
Created method InsertRaw() which inserts a value into dictionary without calling AddToTarget()

Commit:
format

Commit:
Don't Flush() during StringFinalAnalyze but just UpdateMaxUniqueCount, otherwise it messes up the DetectBestCompressionMethod() choice

Commit:
Revert DICT_FSST changes for now, to pass regression_test_storage_size.py

Commit:
Merge remote-tracking branch 'origin/main' into dictcompression_primitivedict

Commit:
Use PrimitiveDictionary instead of std::unordered_map for DICT_FSST compression

Commit:
Remove unnecessary + 1

Commit:
Allocate less for DICT_FSST

Commit:
Refactoring + Possible Dictionary Compression Optimization

Commit:
Remove unnecessary comments, remove InsertRawNoChecks as there's no significant performance gain

lnkuiper reviewed (changes requested):
Nice work! I have a few requests below:

Commit:
Update storage persistent to v1.4.0 on dict compression benchmarks

Commit:
Remove InsertRaw(), use templated Insert instead

Commit:
Deduplicate initialization code by calling Clear()

lnkuiper reviewed (approved):
Thanks for the changes! Great results. LGTM once CI passes

Commit:
Fix compiler error by adding methods to PrimitiveCastOperator

Added WriteSize and WriteToStream to PrimitiveCastOperator.
This error popped up when templating the Insert() function with 	template <bool ADD_TO_TARGET = false>

Commit:
*2 allocation for dictionary
